core/store/src/trie/trie_storage.rs

```diff
@@ -184,30 +245,90 @@ impl TrieCachingStorage {
+            // full cost for it.
+            CachePosition::None => {
+                RawBytesWithCost { value: None, cost: TrieNodeRetrievalCost::Full }
+            }
+            // If value is present only in shard cache, we can copy value to the chunk cache. The cost must be full
+            // anyway because we didn't access the node in the chunk yet.
+            CachePosition::ShardCache(value) => {
+                self.try_put_to_chunk_cache(key.clone(), value.clone());
+                RawBytesWithCost { value: Some(value), cost: TrieNodeRetrievalCost::Full }
+            }
+            // If value is present in chunk cache, the cost is free.
+            CachePosition::ChunkCache(value) => {
+                RawBytesWithCost { value: Some(value), cost: TrieNodeRetrievalCost::Free }
+            }
+        }
+    }
+
+    /// Put a key-value pair into cache.
+    pub(crate) fn put_to_cache(&self, key: CryptoHash, value: Arc<[u8]>) {
+        self.shard_cache.put(key.clone(), value.clone());
+        self.try_put_to_chunk_cache(key, value.clone());
+    }
+
+    /// Set cache state.
+    pub fn set_state(&self, state: TrieCacheState) {
+        self.cache_state.set(state);
+    }
 }
 
 impl TrieStorage for TrieCachingStorage {
     fn retrieve_raw_bytes(&self, hash: &CryptoHash) -> Result<Arc<[u8]>, StorageError> {
-        self.inc_counter();
-        let mut guard = self.cache.0.lock().expect(POISONED_LOCK_ERR);
-        if let Some(val) = guard.get(hash) {
-            Ok(val.clone())
-        } else {
-            let key = Self::get_key_from_shard_uid_and_hash(self.shard_uid, hash);
-            let val = self
-                .store
-                .get(ColState, key.as_ref())
-                .map_err(|_| StorageError::StorageInternalError)?;
-            if let Some(val) = val {
-                let val: Arc<[u8]> = val.into();
-                if val.len() < TRIE_LIMIT_CACHED_VALUE_SIZE {
-                    guard.put(*hash, val.clone());
+        // Try to get value from cache and determine node retrieval cost.
+        let RawBytesWithCost { value, cost } = self.get_from_cache(hash);
+
+        // If full cost for retrieving the node must be charged, increment the counter.
+        if let TrieNodeRetrievalCost::Full = cost {
+            self.inc_counter();
+        }
+
+        match value {
+            // If value is not present, get it from the store and put into the cache.
+            None => {
+                let key = Self::get_key_from_shard_uid_and_hash(self.shard_uid, hash);
+                let val = self
+                    .store
+                    .get(ColState, key.as_ref())
+                    .map_err(|_| StorageError::StorageInternalError)?;
+                if let Some(val) = val {
+                    let val: Arc<[u8]> = val.into();
+                    if val.len() < TRIE_LIMIT_CACHED_VALUE_SIZE {
+                        self.put_to_cache(*hash, val.clone());
+                    }
```

Comment:
This condition feels wrong to me -- it's OK to skip `shard_cache` based on value size, but it's *not* ok to skip the `chunk_hash`. I think the property we want is "adjusting `TRIE_LIMIT_CACHED_VALUE_SIZE` is not  protocol change", but in the current implementation it **would** be one. WE also should have a test for this case.

Reply:
:thinking: yeah, it might be ok to make "values less than X bytes are cached" to be a property of the protocol. However, with the current implementation, we are not looking at the size of values, we are looking at the size of the whole encoding node. That is, we also leak our physical representation of the data on disk... 

That is, if, eg, we remove the `memory_usage` field from `RawTrieNodeWithSize`, that's also going to be a protocol change.

That is, the problem here is that we define the protocol not in terms of abstract shape of the trie (sizes of values or extension keys), but in terms of specific implementation.

Reply:
:thinking: so, we actually **have** to make some such constant a part of the protocol. Originally I was thinking "let's just not cache leaf nodes (values)", but we of course can have an intermediate extension node with 4mb key, and we should have something in the protocol that prevents caching of such a gigantic node.

I feel pretty strongly that the condition for caching should be definable in terms of logical shape of the trie. That is, ideally we *first* decode the node, then look at the combined length of node's key and value, and make caching decision after that. This is in contrast to today's approach where we make caching decision prior to decoding.

However, taking a second look at the current implementation, it does seem that it can be reformulated in terms of logical shape of the trie. I think the rule is:

```
Trie node is cacheable if any of the following holds:

* it's an internal node
* it's a leaf node with key length <= (4000 - 4 - 4 - 4 - 32)
* it's an extension node with key length <= (4000 - 4 - 4 - 32)
* it's a value node with value length <= 4000
```

This seems workable, *but* we need tests for all those cases.